File: caraml-core/caraml-core/CaramlJS.m
//
//  CaramlJS.m
//  caraml-core
//
//  Created by Eric Lange on 1/10/19.
//  Copyright © 2019 LiquidPlayer. All rights reserved.
//

#import "CaramlJS.h"
#import "CaramlJS_private.h"
#import "CaramlView.h"
#import "CaramlView_private.h"

@interface CaramlJS()
@property (nonatomic, strong) CaramlView *m_view;
@end

@implementation CaramlJS
- (id) init:(JSContext*)context view:(CaramlView*)view
{
    self = [super init];
    if (self != nil) {
        self.m_view = view;
    }
    return self;
}

- (void) attach:(NSObject<LCCaramlSurface>*)surface
{
    [self.m_view attach:surface onRestore:NO];
}

- (void) detach
{
    [self.m_view detach];
}

+ (CaramlJS*) from:(JSValue*)value
{
    if ([value.class conformsToProtocol:@protocol(CaramlJSExports)]) {
        return (CaramlJS*) value;
    }
    @throw [NSException exceptionWithName:@"NotACaramlJSObjectException"
                                   reason:@"Object is not a CaramlJS object"
                                 userInfo:nil];
}

@end

File: caraml-core-android/src/main/java/org/liquidplayer/caraml/CaramlJS.java
/*
 * Copyright (c) 2019 Eric Lange
 *
 * Distributed under the MIT License.  See LICENSE.md at
 * https://github.com/LiquidPlayer/caraml-core for terms and conditions.
 */
package org.liquidplayer.caraml;

import android.support.annotation.NonNull;

import org.liquidplayer.javascript.JSContext;
import org.liquidplayer.javascript.JSObject;
import org.liquidplayer.javascript.JSValue;

/**
 * The object passed from JavaScript which provides the attach/detach API.
 */
public class CaramlJS extends JSObject {
    private static class NotACaramlJSObjectException extends RuntimeException {
        NotACaramlJSObjectException() {
            super("Object is not a CaramlJS API object");
        }
    }

    CaramlJS(@NonNull JSContext context, @NonNull CaramlView view) {
        super(context);
        m_view = view;
    }

    /*package*/ void setView(@NonNull CaramlView view) {
        m_view = view;
    }

    /**
     * Attaches a {@code #CaramlSurface} to the view represented by this {@code #CaramlJS} object.
     * @param surface The surface to attach.
     */
    public void attach(@NonNull CaramlSurface surface) {
        m_view.attach(surface,false);
    }

    /**
     * If a surface is attached to the view represented by this {@code #CaramlJS} object, it will
     * be detached, otherwise this will perform no action.
     */
    public void detach() {
        m_view.detach();
    }

    /**
     * Gets the {@code #CaramlJS} object from the JavaScript object returned to node.
     * @param object The JavaScript value passed to the node JS API
     * @return The Java API object
     */
    public static CaramlJS from(JSValue object) {
        if (CaramlJS.class.isAssignableFrom(object.getClass())) {
            return (CaramlJS)object;
        }
        throw new NotACaramlJSObjectException();
    }

    private CaramlView m_view;
}
